src/ow1/heroes/ashe.opy

```diff
@@ -14,4 +16,29 @@ rule "[ashe.opy]: Detect Ashe initialization":
 def initAshe():
     @Name "[ashe.opy]: initAshe()"
     # [TODO]
-    pass
\ No newline at end of file
+    pass
+
+
+rule "[ashe.opy]: Add ultimate charge while using BOB":
+    @Event eachPlayer
+    @Hero ashe
+    @Condition eventPlayer.isUsingUltimate() == true
+
+    eventPlayer.dmg_done_while_ult_active = 0
+    eventPlayer.start_time_ult = getTotalTimeElapsed()
+
+    waitUntil(eventPlayer.isUsingUltimate() == false, 99999)
+    # Check if the user did not switch hero after using B.O.B to add the ult charge
+    if eventPlayer.getCurrentHero() == Hero.ASHE :
+        # Add charge for passive generation and then the ult charge generated from the damage made by Ashe
+        eventPlayer.setUltCharge(percent((ULT_CHARGE_PER_SECOND * (getTotalTimeElapsed() - eventPlayer.start_time_ult) + eventPlayer.dmg_done_while_ult_active) / OW1_ASHE_ULT_COST)) 
+ 
+
+
+rule "[ashe.opy]: Calcul of the damage made while BOB is active":
+    @Event playerDealtDamage
+    @Hero ashe
+    @Condition eventPlayer.isUsingUltimate() == true
+    @Condition victim != attacker
+
+    eventPlayer.dmg_done_while_ult_active += eventDamage
```

Comment:
Can you rewrite this using the `missing_ult_points` variable from `src/ow1/passives/no_ult_negation.opy`?

Your idea is solid, but we were planning to use the logic from that file to fix this issue. This works, but it slightly deviates from the ult charge compensation logic defined in that file.

Reply:
Implemented with `missing_ult_points`